### PR TITLE
ROX-8735: Do not fail on unrecognized resources during deployment check.

### PR DESCRIPTION
**Pull request by dhaus67**
## Description

Previously, we would fail on YAML files during deployment check if they had any objects that are not recognized, i.e. CRDs of operators.
Most prominently people are using `helm template chart > output.yaml` && `roxctl deployment check … -f output.yaml`  for helm charts and this may include multiple resources which we do not yet recognize.

Taking the following YAML as example for such a situation:
```yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  name: frontend
  labels:
    app: nginx
spec:
  selector:
    matchLabels:
      app: nginx
  template:
    metadata:
      labels:
        app: nginx
    spec:
      containers:
      - name: nginx
        image: nginx
---
apiVersion: istio.banzaicloud.io/v1beta1
kind: Istio
metadata:
  name: istio-sample
spec:
  autoInjectionNamespaces:
  - default
  citadel:
    image: 'docker.io/istio/citadel:1.1.2'
    replicaCount: 1
  defaultPodDisruptionBudget:
    enabled: true
  galley:
    image: 'docker.io/istio/galley:1.1.2'
    replicaCount: 1
  gateways:
    egress:
      maxReplicas: 5
      minReplicas: 1
      replicaCount: 1
      sds:
        image: node-agent-k8s
    ingress:
      maxReplicas: 5
      minReplicas: 1
      replicaCount: 1
      sds:
        image: node-agent-k8s
    k8singress: {}
  imageHub: docker.io/istio
  imageTag: 1.1.0
  includeIPRanges: '*'
  mixer:
    image: 'docker.io/istio/mixer:1.1.2'
    maxReplicas: 5
    minReplicas: 1
    replicaCount: 1
  mtls: false
  nodeAgent:
    image: 'docker.io/istio/node-agent-k8s:1.1.2'
  outboundTrafficPolicy:
    mode: ALLOW_ANY
  pilot:
    image: 'docker.io/istio/pilot:1.1.2'
    maxReplicas: 5
    minReplicas: 1
    replicaCount: 1
    traceSampling: 1
  proxy:
    image: 'docker.io/istio/proxyv2:1.1.2'
  proxyInit:
    image: 'docker.io/istio/proxy_init:1.1.2'
  sds: {}
  sidecarInjector:
    image: 'docker.io/istio/sidecar_injector:1.1.2'
    replicaCount: 1
    rewriteAppHTTPProbe: true
  tracing:
    zipkin:
      address: 'zipkin.istio-system:9411'
  version: 1.1.2
```

This would result in the following error when trying to check this deployment:
```shell
$ roxctl deployment check --insecure-skip-tls-verify --endpoint "localhost:8000" -p "<password>" -f tests/yamls/multi-deployment-crd.yaml
ERROR:	checking deployment failed after 3 retries: retrieving alerts from central: could not check deploy-time alerts: rpc error: code = InvalidArgument desc = could not parse YAML: no kind "Istio" is registered for version "istio.banzaicloud.io/v1beta1" in scheme "pkg/runtime/scheme.go:100": invalid arguments
```

Now, instead of returning an error and aborting the deployment check, we should instead ignore unregistered resources, check the resources which we have support of, and give callers feedback which resources were skipped.

With the changes, the deployment check now will return the following with the example YAML:
```bash
$ roxctl deployment check --insecure-skip-tls-verify --endpoint "localhost:8000" -p "<password>" -f tests/yamls/multi-deployment-crd.yaml
INFO:	Ignored object "istio-sample[istio.banzaicloud.io/v1beta1, Kind=Istio]" as its schema was not registered.
Policy check results for deployments: [frontend]
(TOTAL: 6, LOW: 3, MEDIUM: 2, HIGH: 1, CRITICAL: 0)
.
.
.
```

## Checklist
- [X] Investigated and inspected CI test results
- [X] Unit test and regression tests added
- [ ] Evaluated and added CHANGELOG entry if required
- [ ] Determined and documented upgrade steps
- [ ] Documented user facing changes (create PR based on [stackrox/openshift-docs](https://github.com/stackrox/openshift-docs) and merge into [rhacs-docs](https://github.com/openshift/openshift-docs/tree/rhacs-docs))

If any of these don't apply, please comment below.

## Testing Performed
- QA tests via `tests/yamls/`
- Unit tests

### CODE REVIEW

File: central/detection/service/service_impl.go [hunk 1]
```diff
@@ -219,44 +220,68 @@ func (s *serviceImpl) runDeployTimeDetect(ctx context.Context, enrichmentContext
 	return s.enrichAndDetect(ctx, enrichmentContext, deployment, policyCategories...)
 }
 
-func getObjectsFromYAML(yamlString string) ([]k8sRuntime.Object, error) {
+func getObjectsFromYAML(yamlString string) ([]k8sRuntime.Object, []string, error) {
```
Comment (janisz) [inline central/detection/service/service_impl.go:223]: I think we should change name of this function or return a struct with objects and ignore objects. `[]string` tells nothing about what's inside.

Comment (janisz) [reply central/detection/service/service_impl.go:223]: alternatively 
```suggestion
func getObjectsFromYAML(yamlString string) (objects []k8sRuntime.Object, ignoredObjectsRefs []string, err error) {
```

File: central/detection/service/service_impl.go [hunk 2]
```diff
@@ -219,44 +220,68 @@ func (s *serviceImpl) runDeployTimeDetect(ctx context.Context, enrichmentContext
...
 		if err != nil {
-			return nil, errox.InvalidArgs.New("could not parse yaml").CausedBy(err.Error())
+			// Only return errors if the resource's schema is not registered.
+			if !k8sRuntime.IsNotRegisteredError(err) {
+				return nil, nil, errox.InvalidArgs.New("could not parse yaml").CausedBy(err.Error())
+			}
+			// Save the ignored object, so we can return it to the caller and skip it.
+			ignoredObj, err := getIgnoredObjectRefFromYAML(string(yamlBytes))
+			if err != nil {
+				return nil, nil, errox.InvariantViolation.New("could not get ignored object").CausedBy(err.Error())
```
Comment (janisz) [inline central/detection/service/service_impl.go:244]: I think we don't need to extract string for `CausedBy`
```suggestion
				return nil, nil, errox.InvariantViolation.New("could not get ignored object").CausedBy(err)
```

File: central/detection/service/service_impl.go [hunk 3]
```diff
@@ -219,44 +220,68 @@ func (s *serviceImpl) runDeployTimeDetect(ctx context.Context, enrichmentContext
...
 }
 
-func getObjectsFromList(list *coreV1.List) ([]k8sRuntime.Object, error) {
+func getObjectsFromList(list *coreV1.List) ([]k8sRuntime.Object,
+	[]string, error) {
 	objects := make([]k8sRuntime.Object, 0, len(list.Items))
+	var ignoredObjectsRefs []string
 	for i, item := range list.Items {
 		obj, _, err := workloadDeserializer.Decode(item.Raw, nil, nil)
 		if err != nil {
```
Comment (janisz) [inline central/detection/service/service_impl.go:270]: if we invert this condition we can reduce indentation level and make this more readable as we will have short circuit when everything is fine and then not indented ignored objects handling.

File: tests/yamls/multi-deployment-crd.yaml [hunk 4]
```diff
@@ -0,0 +1,80 @@
+apiVersion: apps/v1
```
Comment (janisz) [inline tests/yamls/multi-deployment-crd.yaml:1]: where this file is used?

**Review by janisz — CHANGES_REQUESTED** [review]
Added comments

File: tests/yamls/multi-deployment-crd.yaml (hunk 4 above)
Comment (dhaus67) [reply tests/yamls/multi-deployment-crd.yaml:1]: It'll be called in `roxctl_verification.sh`, which will be called within [run.sh](https://github.com/stackrox/stackrox/blob/6787c87a8ec2f2f409b4cd80a4ca581d73e871e0/tests/e2e/run.sh#L138).
I added these tests to the list of refactoring to bats within [ROX-8801](https://issues.redhat.com/browse/ROX-8801) so we don't forget about them.

Comment (janisz) [reply tests/yamls/multi-deployment-crd.yaml:1]: So maybe add a comment in the file

Comment (dhaus67) [reply tests/yamls/multi-deployment-crd.yaml:3]: Added TODO marker within `roxctl_verification.sh` as well as comment header for all YAML files which resources they contain + that they are being used wtihin `roxctl_verification.sh`.

**Pull request merged**